Spanner/tests/Perf/Operation.php

```diff
@@ -0,0 +1,267 @@
+            }
+        }
+
+        return $this->latency;
+    }
+
+    /**
+     * Executes an operation.
+     *
+     * @param string $database
+     * @param string $table
+     * @param array $operation
+     * @return void
+     */
+    private function runOperation($database, $table, $operation)
+    {
+        $key = $this->keys[array_rand($this->keys)];
+        switch ($operation) {
+            case 'read':
+                $optime = $this->performRead($database, $table, $key);
+                break;
+            case 'update':
+                $optime = $this->update($database, $table, $key);
+                break;
+            case 'insert':
+                $optime = $this->insert($database, $table);
+                break;
+            case 'scan':
+                $optime = $this->scan($database, $table, $key);
+                break;
+        }
+
+        return $optime;
+    }
+
+    /**
+     * Executes a read.
+     *
+     * @param string $database
+     * @param string $table
+     * @param string $key
+     * @return float Time elapsed.
+     */
+    private function performRead($database, $table, $key)
+    {
+        $startTime = microtime(true);
+
+        // Kind of assuming that id is ubiquitous...
+        $results = $database->execute('SELECT * FROM ' . $table . ' where id = @id', [
+            'parameters' => [
+                'id' => $key
+            ]
+        ]);
+
+        foreach ($results as $row) {
+            $key = $row;
+        }
+
+        return microtime(true) - $startTime;
+    }
+
+    /**
+     * Executes a single update.
+     *
+     * @param string $database
+     * @param string $table
+     * @param string $key
+     * @return float Time elapsed.
+     */
+    private function update($database, $table, $key)
+    {
+        // Does a single update operation.
+        $field = rand(0, 9);
+        $startTime = microtime(true);
+
+        $operation = $database->transaction(['singleUse' => false])
+            ->update($table, [
+                'id' => $key,
+                'field' . $field => $this->randString(false, 100)
+            ])->commit();
+
+        return microtime(true) - $startTime;
+    }
+
+    /**
+     * Inserts a row
+     *
+     * @param string $database
+     * @param string $table
+     * @return float Time elapsed.
+     */
+    private function insert($database, $table)
+    {
+        $batch = [];  //array of $arrFields
+        $fields = [];
+        $fields['id'] = 'user4' . $this->randString(true, 17);
+
+        for ($f = 0; $f < 10; $f++) {
+            $fields['field' . $f] = $this->randString(false, 100);
+        }
+
+        $batch[] = $fields;
+        array_multisort($batch);
+
+        $startTime = microtime(true);
+        $operation = $database->transaction(['singleUse' => true])
+            ->insertBatch($table, $batch)
+            ->commit();
+
+        return microtime(true) - $startTime;
+    }
+
+    /**
+     * Not Implemented.
```

Comment:
Is there any need to have this method then?

Reply:
My thought was, as "scan" is a part of YCSB, it would be somewhat odd to omit it entirely. Additionally, the function was provided and empty in the initial implementation.

Would raising an E_USER_NOTICE be good? "Your workload contains scan operations, but scans are not currently implemented"?

Reply:
Yeah and I would just put it in the switch statement. We aren't fulfilling a contract here so I don't see the need for any dead code